Add smoke-tests runner and separate util functions

sstamenk opened the pull request:
## Motivation

Fixes segmentation faults during smoke-tests execution reported in [#2325](https://github.com/ROCm/TheRock/issues/2325)
## Technical Details

Added a launch script for smoke-tests similar to run_linux_pytorch_tests.py. The script reuses logic from run_linux_pytorch_tests.py to set HIP_VISIBLE_DEVICES to the proper device index to avoid segmentation faults. 

Extracted common GPU query logic to the pytorch_utils.py file.
Combined querying of visible and supported AMDGPU archs into a single subprocess call.

## Test Plan

Running smoke-tests.

## Test Result

Tests passing successfully without segfaults.

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

ScottTodd commented:
I ran out of time today to review these PRs (still 11 PRs in my queue x_x). @sstamenk , can you and @lamikr discuss a bit / review each other's PRs / come up with a plan together?

lamikr commented:
I am traveling/holiday in this week so my time is quite limited especially today. Couple of things:

1) We both seems to have idea of running a launcher to launch the smoketests.

2) We both have single subprocess call to detect which GPU's to use. The difference is that I combined two methods to single one. In principle both ways are ok for me and I don't have strong opinion.

3) Biggest difference is in the approach whether we want to give the tested GPUs target family as a parameter and use that information for querying the target devices. That's here in pytorch_utils.py with 3 different cases starting from line 269. That code would run the tests for first detected GPUs parameter is not passed.

In my code I do not have that code at all and instead run the tests for all supported GPUs. So there is no need to check the AMDGPU_FAMILY environment variable and do the checking for whether start of it's name matches with the individual GPUs gfx-name. In addition I don't need to have the gpu-index remapping code that's on same area pytorch_utils.py.

We can query from the pytorch directly the list of GPUs that it has build in supports. Then we can unset the HIP_VISIBLE_DEVICES and ask what are the GPU's installed and then construct a new list which contains GPU index entries that were in both lists. 

Example 1: We have pytorch build with gfx110X-all as a target.  When we query from the pytorch the supported GPUs, we would get a list [gfx1100, gfx1101, gfx1102, gfx1103]  
Then lets assume that query of installed GPUs would return: [gfx1201, gfx1103, [gfx1102, gfx1103]
--> gfx1102 and gfx1103 would be the GPUs we can test with this pytorch. Id list we return [0, 1]
--> When looping the ID list, we would set HIP_VISIBLE_DEVICES=0 and  HIP_VISIBLE_DEVICES=0 while running the tests

Example 2: We have pytorch build with gfx103X-all as a target.  When we query from the pytorch the supported GPUs, we could get a list [gfx1030, gfx1031, gfx1032, gfx1035]  
Then lets assume that the query of installed GPUs would return: [gfx1201, gfx1030]
--> gfx1030 would be the only GPU we can test with this pytorch. Id list we return [1]
--> When looping the ID list, we would set HIP_VISIBLE_DEVICES=1 while running the tests

4) This code tries to unify some code with the another test runner under build_tools folder. That is good. I did not have that done in 2457. Maybe we could somehow combine these approaches.

sstamenk commented:
`AMDGPU_FAMILY `and `amdgpu_family` are both optional as you pointed out, if they are left out then the filtering is done the same way as your proposed solution. I guess we can remove these first 2 cases since they aren't used but I would like @ScottTodd input.

I agree we should run it on all supported GPUs instead of the first one only since for example gfx120x would support both gfx1201 and gfx1200, but filter to only run on unique devices (so we dont run multiple times on the same architecture).

@ScottTodd in the future will the wheels be sperate for each architecture or will they remain aggregated like now or perhaps even further unified (single wheel for all Radeon architectures)? As is, the way I understand it, is that we could have a gfx1200 specific issue that isn't caught by these tests because we tested a gfx120x wheel on a node that has only gfx1201. 
What's the better approach here? Will there be a separate workflow job for each of the architectures?

We can do the logic to run on all supported GPUs instead of only the first but that still wouldn't guarantee that we test a wheel an AMDGPU family on all of the specific architectures it supports (for example the node that originally caused the issue had gfx1100 and gf1201 but was testing gfx120x, so gfx1200 wouldn't have been tested on even if we did run on all supported devices). How are the nodes where the workflow job is done selected?

ScottTodd commented on external-builds/pytorch/run_pytorch_smoke_tests.py:
```diff
@@ -0,0 +1,92 @@
+        pytorch_args += [
+            "-p",
+            "no:cacheprovider",  # Disable caching: useful when running in a container
+        ]
```
Let's call this variable `pytest_args`, not `pytorch_args`.

sstamenk replied:
Done

sstamenk commented on external-builds/pytorch/pytorch_utils.py:
```diff
@@ -0,0 +1,347 @@
+    """Detect and configure AMDGPU family for testing.
+
+    This function queries available GPUs and sets HIP_VISIBLE_DEVICES BEFORE
+    PyTorch/CUDA is initialized in the main process via pytest.
```
Reworked the function not to set HIP_VISIBLE_DEVICES, instead this is done in the runner scripts explicitly. This way the function has no side effects (at least related to the setting of env variables).

sstamenk commented:
> Thanks, I like how the code is reused here.
> 
> We might be able to further simplify by setting the `HIP_VISIBLE_DEVICES` environment variable more consistently in the test workflows or on the test runners themselves.

Reworked the functions to only provide a dictionary with (gpu_arch: str, device_index: list[int]) (key, value) pairs. Also added a function that returns only unique devices by taking just the first index. These functions no longer set HIP_VISIBLE_DEVICES and this is instead done in the runner scripts instead.

Smoke-tests are now executed on every supported unique arch while PyTorch unit tests are only done on the first unique supported device.

sstamenk commented on external-builds/pytorch/pytorch_utils.py:
```diff
@@ -0,0 +1,346 @@
+        sys.exit(1)
+    except Exception as e:
+        print(f"\n[ERROR] Unexpected error retrieving available GPUs: {e}")
+        sys.exit(1)
```
Removed these as they were unused.

ScottTodd commented on external-builds/pytorch/run_pytorch_tests.py:
```diff
@@ -218,8 +218,12 @@ def main() -> int:
+    # For unit tests, use only the first supported device
+    first_arch, first_device_idx = next(iter(unique_supported_devices.items()))
+    os.environ["HIP_VISIBLE_DEVICES"] = str(first_device_idx)
+    print(f"Using AMDGPU family: {first_arch} (device index: {first_device_idx})")
```
We're going to need to iterate on this more as we add more test coverage including mult-gpu tests (https://github.com/ROCm/TheRock/issues/2507) and we're likely going to switch from directly invoking pytest to running the `run_test.py` file provided by pytorch (see https://github.com/ROCm/TheRock/issues/2406).

I think this is okay for now as we can relatively easily change the behavior here.

cc @jithunnair-amd @pruthvistony

sstamenk replied:
Had an idea to encapsulate this behavior in a function. set_gpu_execution_policy now sets HIP_VISIBLE_DEVICES according to one of 4 policies (single, unique-single, unique, all). This is called from the runner scripts main function with an appropriate policy ("single" for PyTorch unit-tests, iterates over "unique-single" for smoke-tests to run on each unique device).

ScottTodd commented:
Ready to merge, thanks (was expecting you as the author to merge).
